tests/common/gcp_type/firewall_rule_test.py

```diff
@@ -0,0 +1,433 @@
+                'firewall_rule_direction': 'egress',
+                'firewall_rule_network': 'n1',
+            },
+            {
+                'firewall_rule_destination_ranges': json.dumps(['10.0.0.1',
+                                                                '10.0.0.2']),
+                'firewall_rule_direction': 'egress',
+                'firewall_rule_network': 'n1',
+            },
+            True,
+        ),
+    ])
+    def test_firewall_rule_gt(self, rule_1_dict, rule_2_dict, expected):
+        """Tests that rule 1 > rule 2 returns the correct value."""
+        rule_1 = firewall_rule.FirewallRule(**rule_1_dict)
+        rule_2 = firewall_rule.FirewallRule(**rule_2_dict)
+        self.assertEqual(expected, rule_1 > rule_2)
+
+    @parameterized.parameterized.expand([
+        (
+            {
+                'firewall_rule_source_ranges': json.dumps(['0.0.0.0/0']),
+                'firewall_rule_direction': 'ingress',
+                'firewall_rule_network': 'n1',
+                'firewall_rule_allowed': json.dumps(
+                    [{'IPProtocol': 'tcp', 'ports': ['22']}]),
+            },
+            {
+                'firewall_rule_source_ranges': json.dumps(['0.0.0.0/0']),
+                'firewall_rule_direction': 'ingress',
+                'firewall_rule_network': 'n1',
+                'firewall_rule_allowed': json.dumps(
+                    [{'IPProtocol': 'tcp', 'ports': ['22']}]),
+            },
+            True,
+        ),
+        (
+            {
+                'firewall_rule_source_ranges': json.dumps(['0.0.0.0/0']),
+                'firewall_rule_direction': 'ingress',
+                'firewall_rule_network': 'n1',
+            },
+            {
+                'firewall_rule_source_ranges': json.dumps(['1.1.1.1']),
+                'firewall_rule_direction': 'ingress',
+                'firewall_rule_network': 'n1',
+            },
+            False,
+        ),
+    ])
+    def test_firewall_rule_eq(self, rule_1_dict, rule_2_dict, expected):
+        """Tests that rule 1 == rule 2 returns the correct value."""
+        rule_1 = firewall_rule.FirewallRule(**rule_1_dict)
+        rule_2 = firewall_rule.FirewallRule(**rule_2_dict)
+        self.assertEqual(expected, rule_1 == rule_2)
+
+    @parameterized.parameterized.expand([
+        (
+            {
+                'firewall_rule_allowed': json.dumps(
+                    [{'IPProtocol': 'tcp', 'ports': ['22']}]),
+                'firewall_rule_denied': json.dumps(
+                    [{'IPProtocol': 'tcp', 'ports': ['22']}]),
+            },
+        ),
+    ])
+    def test_firewall_action_init_error(self, action_1_dict):
+        """Tests that initializing a bad FirewallAction raises an error."""
+        with self.assertRaises(ValueError):
+            _ = firewall_rule.FirewallAction(**action_1_dict)
+
+    @parameterized.parameterized.expand([
```

Comment:
I recommend having one or more parameterized tests where there are multiple protocols and ports listed in a rule. Include them in different orders in self and other to ensure sorting works correctly. Include ports in different orders as well.